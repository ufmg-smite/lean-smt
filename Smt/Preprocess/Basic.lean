/-
Copyright (c) 2021-2024 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Abdalrhman Mohamed
-/

import Lean.Meta.Basic
import Lean.Meta.InferType

namespace Smt.Preprocess

open Lean

structure Result where
  map : Std.HashMap Expr (Array Expr)
  hs : Array Expr
  mv : MVarId

/-- Return all propositions in the local context. -/
def getPropHyps : MetaM (Array FVarId) := do
  let mut result := #[]
  for localDecl in (← getLCtx) do
    unless localDecl.isImplementationDetail do
      if ← pure !(isNonEmpty localDecl.type) <&&> Meta.isProp localDecl.type then
        result := result.push localDecl.fvarId
  return result
where
  isNonEmpty (e : Expr) : Bool :=
  match_expr e with
  | Nonempty _ => true
  | _ => false

end Smt.Preprocess
