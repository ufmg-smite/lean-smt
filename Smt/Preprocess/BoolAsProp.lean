import Smt.Reconstruct.Prop.Core

theorem Bool.not_eq_true2 (x : Bool) : (!x : Bool) = ¬(x : Prop) := by
  rw [Bool.not_eq_eq_eq_not, Bool.not_true, Bool.not_eq_true]

theorem Bool.iff_eq_true (x y : Bool) : (x ↔ y : Bool) = ((x : Prop) ↔ (y : Prop)) := by
  simp only [Bool.coe_iff_coe, decide_eq_true_eq]

theorem Bool.xor_eq_true (x y : Bool) : (x ^^ y : Bool) = (XOr (x : Prop) (y : Prop)) := by
  cases x <;> cases y <;> decide

theorem Bool.eq_eq_true (x y : Bool) : (x = y : Bool) = ((x : Prop) = (y : Prop)) := by
  simp only [decide_eq_true_eq, _root_.eq_iff_iff, coe_iff_coe]

theorem Bool.eq_self (x : Bool): (x = x) = (True : Prop) := by
  simp only

theorem Prop.eq_true (x : Prop) : (x = True : Prop) = (x : Prop) := by
  simp only [eq_iff_iff, iff_true]

theorem Prop.eq_false (x : Prop) : (x = False : Prop) = (¬x : Prop) := by
  simp only [eq_iff_iff, iff_false]
