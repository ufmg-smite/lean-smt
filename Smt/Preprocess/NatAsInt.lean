theorem Nat.forall_as_int (P : Nat → Prop) : (∀ x : Nat, P x) ↔ (∀ x : Int, x ≥ 0 → P x.toNat) := by
  constructor
  · intro h x x_nonneg
    let x_nat : Nat := x.toNat
    have h_nat : x = x_nat := by
      exact Int.eq_natCast_toNat.mpr x_nonneg
    rw [h_nat]
    exact h x_nat
  · intro h x
    let x_int : Int := x
    apply h x_int
    exact Int.ofNat_zero_le x

theorem Nat.exists_as_int (P : Nat → Prop) : (∃ x : Nat, P x) ↔ (∃ x : Int, x ≥ 0 ∧ (x ≥ 0 → P x.toNat)) := by
  constructor
  · intro h
    obtain ⟨x, hx⟩ := h
    exists x
    exact if_false_right.mp fun a => hx
  · intro h
    obtain ⟨x, hx⟩ := h
    let x_nat : Nat := x.toNat
    have h_nat : x = x_nat := by
      refine Int.eq_natCast_toNat.mpr ?_
      exact hx.1
    exists x_nat
    unfold x_nat
    have : x ≥ 0 := hx.1
    exact hx.2 this

def Int.natSub (x y : Int) :=
  if x ≥ y then x - y else 0

theorem Int.natCast_sub2 (x y : Nat) : (x - y : Nat) = Int.natSub x y := by
  rw [Int.natSub]
  split <;> rename_i h
  · rw [Int.natCast_sub]
    exact Int.ofNat_le.mp h
  · rw [Nat.sub_eq_zero_of_le (Nat.le_of_succ_le (Int.ofNat_lt.mp (Int.not_le.mp h))), Int.natCast_zero]

theorem Int.ofNat_eq (x y : Nat) : (x = y) = ((x : Int) = (y : Int)) := by
  simp only [Int.ofNat_inj]

theorem Int.ofNat_le2 (x y : Nat) : (x ≤ y) = ((x : Int) ≤ (y : Int)) := by
  simp only [Int.ofNat_le]

theorem Int.ofNat_lt2 (x y : Nat) : (x < y) = ((x : Int) < (y : Int)) := by
  simp only [Int.ofNat_lt]

theorem Int.ofNat_ge (x y : Nat) : (x ≥ y) = ((x : Int) ≥ (y : Int)) := by
  simp only [ge_iff_le, Int.ofNat_le]

theorem Int.ofNat_gt (x y : Nat) : (x > y) = ((x : Int) > (y : Int)) := by
  simp only [gt_iff_lt, Int.ofNat_lt]

theorem Int.ofNat_ne (x y : Nat) : x ≠ y ↔ ((x : Int) ≠ (y : Int)) := by
  rw [← Decidable.not_iff_not]
  rw [Decidable.not_not, Decidable.not_not]
  rw [Int.ofNat_inj]

theorem doubleCast (p : Nat → Prop) (w : Nat) : p w = p (w : Int).toNat := by
  rw [← congrArg p (Int.toNat_natCast w)]
