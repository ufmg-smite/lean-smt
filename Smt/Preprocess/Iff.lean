/-
Copyright (c) 2021-2024 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Abdalrhman Mohamed, Tomaz Gomes Mascarenhas
-/

import Smt.Preprocess.Basic
import Qq

namespace Smt.Preprocess

open Lean Qq

theorem iff_eq_eq : (p ↔ q) = (p = q) := propext ⟨propext, (· ▸ ⟨(·), (·)⟩)⟩

theorem eq_resolve {p q : Prop} (hp : p) (hpq : p = q) : q := hpq ▸ hp

def replaceIff (e : Expr) : MetaM Expr :=
  let f e :=
    if let some ((l : Q(Prop)), (r : Q(Prop))) := e.app2? ``Iff then
      q($l = $r)
    else
      none
  Meta.mkAppM ``Eq #[e, e.replace f]

def containsIff (e : Expr) : Bool :=
  (Expr.const ``Iff []).occurs e

def elimIff (mv : MVarId) (hs : Array Expr) : MetaM Result := mv.withContext do
  let t ← instantiateMVars (← mv.getType)
  let ts ← hs.mapM (Meta.inferType · >>= instantiateMVars)
  if !(containsIff t || ts.any containsIff) then
    return { map := Std.HashMap.insertMany ∅ (hs.zip (hs.map .singleton)), hs, mv }
  let simpTheorems ← #[``eq_self, ``iff_eq_eq].foldlM (·.addConst ·) ({} : Meta.SimpTheorems)
  let simpTheorems := #[simpTheorems]
  let congrTheorems ← Meta.getSimpCongrTheorems
  let ctx ← Meta.Simp.mkContext {} simpTheorems congrTheorems
  let (hs', mv') ← elimIffLocalDecls mv hs.toList ctx #[]
  let mv' ← elimIffTarget mv' ctx
  return { map := Std.HashMap.insertMany ∅ (hs'.zip (hs.map .singleton)), hs := hs', mv := mv' }
where
  elimIffLocalDecls mv hs ctx hs' := do match hs with
    | [] => return (hs', mv)
    | h :: hs =>
      let type ← Meta.inferType h
      let eq ← replaceIff (← instantiateMVars type)
      let (_, l, r) := eq.eq?.get!
      if l == r then
        elimIffLocalDecls mv hs ctx (hs'.push h)
      else
        let (res, _) ← Meta.simp eq ctx
        let h' := mkApp4 (.const ``eq_resolve []) l r h (mkOfEqTrue eq (← res.getProof))
        if let .some fv := h.fvarId? then
          let res ← mv.replace fv h' (.some r)
          let hs' := hs'.map res.subst.apply
          let hs := hs.map res.subst.apply
          res.mvarId.withContext (elimIffLocalDecls res.mvarId hs ctx (hs'.push (.fvar res.fvarId)))
        else
          elimIffLocalDecls mv hs ctx (hs'.push h')
      termination_by hs.length
  elimIffTarget mv ctx := mv.withContext do
    let eq ← replaceIff (← instantiateMVars (← mv.getType))
    let (res, _) ← Meta.simp eq ctx
    if res.expr.isTrue then
      mv.replaceTargetEq eq.appArg! (mkOfEqTrue eq (← res.getProof))
    else
      return mv
  mkOfEqTrue p hpt :=
    mkApp2 (.const ``of_eq_true []) p hpt

end Smt.Preprocess
