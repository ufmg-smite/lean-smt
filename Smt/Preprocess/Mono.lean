/-
Copyright (c) 2021-2024 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Abdalrhman Mohamed
-/

import Auto.Tactic
import Smt.Preprocess.Basic

namespace Auto

open Lean Elab Embedding.Lam

def DTr.contains (self : DTr) (other : DTr) : Bool :=
  if self == other then true
  else match self with
    | .leaf _ => false
    | .node _ dtrs => dtrs.attach.any fun ⟨dtr, _⟩ => dtr.contains other

structure InputHints' where
  lemmas   : Array Lemma := #[]
  lemdbs   : Array Name  := #[]
  lctxhyps : Bool        := false
deriving Inhabited, BEq

/--
  From a user-provided term `stx`, produce a lemma
-/
def Prep.elabLemma' (e : Expr) (deriv : DTr) : MetaM Lemma := do
  -- abstract any remaining mvars:
  let e ← instantiateMVars e
  let abstres ← Auto.abstractMVars e
  let e := abstres.expr
  let paramNames := abstres.paramNames
  return Lemma.mk ⟨e, ← Meta.inferType e, deriv⟩ paramNames

def Prep.elabDefEq' (name : Name) : MetaM (Array Lemma) := do
  match (← getEnv).find? name with
  | some (.recInfo val) =>
    -- Generate definitional equation for recursor
    addRecAsLemma val
  | some (.defnInfo val) =>
    -- Generate definitional equation for (possibly recursive) declaration
    match ← Meta.getEqnsFor? name with
    | some eqns => eqns.mapIdxM fun i eq => do
      -- TODO: create mvs for implicit arguments as well.
      let us ← Meta.mkFreshLevelMVarsFor (.defnInfo val)
      elabLemma' (.const eq us) (.leaf s!"defeq {i} {name}")
    | none => return #[]
  | some (.axiomInfo _)  => return #[]
  | some (.thmInfo _)    => return #[]
  -- If we have inductively defined propositions, we might
  --   need to add constructors as lemmas
  | some (.ctorInfo _)   => return #[]
  | some (.opaqueInfo _) => throwError "{decl_name%} :: Opaque constants cannot be provided as lemmas"
  | some (.quotInfo _)   => throwError "{decl_name%} :: Quotient constants cannot be provided as lemmas"
  | some (.inductInfo _) => throwError "{decl_name%} :: Inductive types cannot be provided as lemmas"
  | none => throwError "{decl_name%} :: Unknown constant {name}"

def collectUserLemmas' (exprs : Array Expr) : MetaM (Array Lemma) :=
  Meta.withNewMCtxDepth do
    let mut lemmas := #[]
    for expr in exprs do
      let ⟨⟨proof, type, deriv⟩, params⟩ ← Prep.elabLemma' expr (.leaf s!"❰{expr}❱")
      if ← Prep.isNonemptyInhabited type then
        throwError "invalid lemma {type}, lemmas should not be inhabitation facts"
      else if ← Meta.isProp type then
        lemmas := lemmas.push ⟨⟨proof, ← instantiateMVars type, deriv⟩, params⟩
      else
        -- **TODO**: Relax condition?
        throwError "invalid lemma {type} for auto, proposition expected"
    return lemmas

def collectHintDBLemmas' (names : Array Name) : MetaM (Array Lemma) := do
  let mut hs : Std.HashSet Name := Std.HashSet.emptyWithCapacity
  let mut ret : Array Lemma := #[]
  for name in names do
    let .some db ← findLemDB name
      | throwError "unknown lemma database {name}"
    let lemNames ← db.toHashSet
    for lname in lemNames do
      if !hs.contains lname then
        hs := hs.insert lname
        ret := ret.push (← Lemma.ofConst lname (.leaf s!"lemdb {name} {lname}"))
  return ret

def collectDefeqLemmas' (names : Array Name) : MetaM (Array Lemma) :=
  Meta.withNewMCtxDepth do
    let lemmas ← names.flatMapM Prep.elabDefEq'
    lemmas.mapM (fun (⟨⟨proof, type, deriv⟩, params⟩ : Lemma) => do
      let type ← instantiateMVars type
      return ⟨⟨proof, type, deriv⟩, params⟩)

def checkDuplicatedFact' (exprs : Array Lemma) : MetaM Unit :=
  let n := exprs.size
  for i in [0:n] do
    for j in [i+1:n] do
      if exprs[i]? == exprs[j]? then
        throwError "Auto does not accept duplicated input terms"

def checkDuplicatedLemmaDB' (names : Array Name) : MetaM Unit :=
  let n := names.size
  for i in [0:n] do
    for j in [i+1:n] do
      if names[i]? == names[j]? then
        throwError "Auto does not accept duplicated lemma databases"

def collectAllLemmas' (inputHints : InputHints') (unfoldInfos : Array Prep.ConstUnfoldInfo)
  (defeqNames : Array Name) (ngoalAndBinders : Array FVarId) :
  -- The first `Array Lemma` are `Prop` lemmas
  -- The second `Array Lemma` are Inhabitation facts
  MetaM (Array Lemma × Array Lemma) := do
  let unfoldInfos ← Prep.topoSortUnfolds unfoldInfos
  let startTime ← IO.monoMsNow
  let lctxLemmas ← collectLctxLemmas inputHints.lctxhyps ngoalAndBinders
  let lctxLemmas ← lctxLemmas.mapM (unfoldConstAndPreprocessLemma unfoldInfos)
  traceLemmas `auto.printLemmas "Lemmas collected from local context:" lctxLemmas
  checkDuplicatedFact' inputHints.lemmas
  checkDuplicatedLemmaDB' inputHints.lemdbs
  let userLemmas := inputHints.lemmas ++ (← collectHintDBLemmas' inputHints.lemdbs)
  let userLemmas ← userLemmas.mapM (unfoldConstAndPreprocessLemma unfoldInfos)
  traceLemmas `auto.printLemmas "Lemmas collected from user-provided terms:" userLemmas
  let defeqLemmas ← collectDefeqLemmas' defeqNames
  let defeqLemmas ← defeqLemmas.mapM (unfoldConstAndprepReduceDefeq unfoldInfos)
  traceLemmas `auto.printLemmas "Lemmas collected from user-provided defeq hints:" defeqLemmas
  trace[auto.tactic] "Preprocessing took {(← IO.monoMsNow) - startTime}ms"
  let inhFacts ← Inhabitation.getInhFactsFromLCtx
  let inhFacts ← inhFacts.mapM (unfoldConstAndPreprocessLemma unfoldInfos)
  traceLemmas `auto.printLemmas "Inhabitation lemmas :" inhFacts
  return (lctxLemmas ++ userLemmas ++ defeqLemmas, inhFacts)

def createInhHyps : FVarId × DTr → MetaM (Option Meta.Hypothesis) := fun (fv, dtr) => do
  let s := match dtr with
    | .leaf s => s
    | .node s _ => s
  if s.startsWith "nonemptyOfAtom" || s.startsWith "lctxInh" then
    let userName ← mkFreshUserName `inst
    let type ← Meta.mkAppM ``Nonempty #[← fv.getType]
    let value ← Meta.mkAppOptM ``Nonempty.intro #[← fv.getType, Expr.fvar fv]
    return some { userName, type, value }
  return none

def mono' (declName? : Option Name) (mv : MVarId) (hints : InputHints') (unfoldInfos : Array Prep.ConstUnfoldInfo) (defeqNames : Array Name) := do
  let (goalBinders, newGoal) ← mv.intros
  let [nngoal] ← newGoal.apply (.const ``Classical.byContradiction [])
    | throwError "{declName?} :: Unexpected result after applying Classical.byContradiction"
  let (ngoal, absurd) ← MVarId.intro1 nngoal
  absurd.withContext do
    let (lemmas, inhFacts) ← collectAllLemmas' hints unfoldInfos defeqNames (goalBinders.push ngoal)
    let (proof, (mvarId, dtrs)) ← runMono declName? lemmas inhFacts
    mvarId.withContext do
      let hs ← dtrs.filterMapM createInhHyps
      let f (acc : MessageData) (dtr : FVarId × DTr) :=
        acc ++ m!"\n<{Expr.fvar dtr.fst}> = <{ToString.toString dtr.snd}>"
      let message : MessageData := dtrs.foldl f m!""
      trace[smt.preprocess] "dtrs: {message}"
      let (_, mv) ← mvarId.assertHypotheses hs
      absurd.assign proof
      return (mv, dtrs)

namespace Tactic

/-- Parse `hints` to an array of `Term`, which is still syntax -/
def parseHints' : TSyntax ``hints → Tactic.TacticM InputHints'
| `(hints| [ $[$hs],* ]) => do
  let mut exprs := #[]
  let mut lemdbs := #[]
  let mut lctxhyps := false
  let elems ← hs.mapM parseHintElem
  for elem in elems do
    match elem with
    | .term t => exprs := exprs.push (← Prep.elabLemma t (.leaf s!"❰{t}❱"))
    | .lemdb db => lemdbs := lemdbs.push db
    | .lctxhyps => lctxhyps := true
  return ⟨exprs, lemdbs, lctxhyps⟩
| `(hints| ) => return ⟨#[], #[], true⟩
| _ => throwUnsupportedSyntax

syntax (name := mono') "mono' " hints (uord)* : tactic

open Lean Elab Tactic in
@[tactic mono'] def evalMono' : Tactic
  | `(mono' | mono' $hints $[$uords]*) => withMainContext do
    let hints ← parseHints' hints
    let (unfoldInfos, defeqNames) ← parseUOrDs uords
    let (mv, _) ← Auto.mono' (← Elab.Term.getDeclName?) (← Tactic.getMainGoal) hints unfoldInfos defeqNames
    Tactic.replaceMainGoal [mv]
  | _ => throwUnsupportedSyntax

end Auto.Tactic

namespace Smt.Preprocess

open Lean

open Auto in
def hintsToAutoHints (hs : Array Expr) : MetaM (Std.HashMap DTr Expr × InputHints' × Array Prep.ConstUnfoldInfo × Array Name) := do
  let lemmas ← hs.mapM inferLemma
  let map := .ofList (lemmas.map (fun l => (l.deriv, l.proof))).toList
  return (map, ⟨lemmas, #[], false⟩, #[], #[])
where
  inferLemma (e : Expr) : MetaM Lemma := do
    let paramNames ← getLevelParamNames e
    return Lemma.mk ⟨e, ← Meta.inferType e, (.node "smtHint" #[.leaf s!"{e}"])⟩ paramNames
  getLevelParamNames {ω : Type} {m} [STWorld ω m] [MonadLiftT (ST ω) m] [Monad m] (e : Expr) : m (Array Name) := do
    let ((), names) ← (e.forEach getLevelParamName).run (m := m) {}
    return names.toArray
  getLevelParamName {m} [Monad m] (e : Expr) : StateT (Std.HashSet Name) m Unit := do
    match e with
    | .const _ lvls =>
      let names := lvls.filterMap fun lvl => do
        let .param name := lvl | none
        some name
      modify fun s => s.insertMany names
    | _ => return

def mono (mv : MVarId) (hs : Array Expr) : MetaM Result := do
  let (invMap, hints, unfoldInfos, defeqNames) ← hintsToAutoHints hs
  let (mv, dtrs) ← Auto.mono' `smt mv hints unfoldInfos defeqNames
  let map := dtrs.foldl (init := {}) fun map (fv, dtr) =>
    let usedHints := invMap.filter (fun k _ => dtr.contains k)
    map.insert (.fvar fv) usedHints.valuesArray
  let hs ← mv.withContext (return (← getPropHyps).map Expr.fvar)
  trace[smt.preprocess] "goal: {mv}"
  return { map, hs, mv }

end Smt.Preprocess
