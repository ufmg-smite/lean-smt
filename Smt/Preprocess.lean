/-
Copyright (c) 2021-2024 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Abdalrhman Mohamed, Tomaz Gomes Mascarenhas
-/

import Smt.Preprocess.Iff
import Smt.Preprocess.Mono
import Smt.Preprocess.PushHintsToCtx
