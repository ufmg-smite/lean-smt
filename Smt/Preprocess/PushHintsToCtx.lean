/-
Copyright (c) 2021-2024 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Abdalrhman Mohamed
-/

import Smt.Preprocess.Basic
import Lean.Meta.Tactic.Assert

namespace Smt.Preprocess

open Lean

def pushHintsToCtx (mv : MVarId) (hs : Array Expr) : MetaM Result := do
  hs.foldrM pushHint { map := {}, hs := #[], mv }
where
  pushHint (h : Expr) (r : Result) : MetaM Result := do
    if h.isFVar || h.isConst then
      return { r with map := r.map.insert h #[h], hs := r.hs.push h }
    else
      let mv' ← r.mv.assert (← mkFreshId) (← Meta.inferType h) h
      let ⟨fv, mv'⟩ ← mv'.intro1
      let h' := .fvar fv
      return { map := r.map.insert h' #[h], hs := r.hs.push h', mv := mv' }

end Smt.Preprocess
