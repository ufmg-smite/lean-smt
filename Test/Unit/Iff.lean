import Smt.Preprocess.Iff

syntax (name := elimIff) "elim_iff " "[" term,* "]" : tactic

open Lean.Elab Tactic in
@[tactic elimIff] def evalReconstruct : Tactic
  | `(tactic| elim_iff [$hs,*]) => withMainContext do
    let mv ← getMainGoal
    let hs ← hs.getElems.mapM (Term.elabTerm · none)
    Lean.logInfo m!"Before: {hs}"
    let ⟨_, hs, mv⟩ ← Smt.Preprocess.elimIff mv hs
    mv.withContext (Lean.logInfo m!"After: {hs}")
    replaceMainGoal [mv]
  | _ => throwUnsupportedSyntax

def t (p : Prop) : p ↔ p := by simp

example (p q r : Prop) (ht : True) (hpqr : (p ↔ q) ∧ (q ↔ r)) (ht' : True) : p ↔ r := by
  elim_iff [hpqr, ht, ht', t]
  simp_all
