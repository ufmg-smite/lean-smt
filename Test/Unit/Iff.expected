Before: [hpqr, ht, ht', t]
After: [hpqr,
 ht,
 ht',
 Smt.Preprocess.eq_resolve t
   (of_eq_true
     (Eq.trans
       (congr (congrArg Eq (forall_congr fun p => Eq.trans Smt.Preprocess.iff_eq_eq (eq_self p)))
         (forall_congr fun p => eq_self p))
       (eq_self (âˆ€ (p : Prop), True))))]
Test/Unit/Iff.lean:18:24: warning: unused variable `ht`
note: this linter can be disabled with `set_option linter.unusedVariables false`
Test/Unit/Iff.lean:18:63: warning: unused variable `ht'`
note: this linter can be disabled with `set_option linter.unusedVariables false`
